raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,149 @@
+ */
+package org.apache.kafka.raft;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.generated.QuorumStateData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.OptionalInt;
+
+/**
+ * Local file based quorum state store.
+ */
+public class FileBasedStateStore implements QuorumStateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(FileBasedStateStore.class);
+
+    static final int UNDEFINED_EPOCH = -1;
+    private static final int UNKNOWN_LEADER_ID = -1;
+    private static final int NOT_VOTED = -1;
+
+    private final File stateFile;
+
+    private final QuorumStateData cachedState;
+
+    public FileBasedStateStore(final File stateFile) throws IOException {
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile);
+        LOG.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file) throws IOException {
+        try (final BufferedReader reader = Files.newBufferedReader(file.toPath())) {
+            final ObjectMapper objectMapper = new ObjectMapper();
+            final short version = readShort(reader);
+            final JsonNode jsonNode = objectMapper.readTree(readLine(reader));
+
+            QuorumStateData data = new QuorumStateData();
+            data.fromJson(jsonNode, version);
+            return data;
+        } catch (final NoSuchFileException e) {
+            return new QuorumStateData();
+        } catch (final EOFException e) {
+            delete();
+            LOG.info("Local quorum file corrupted, deleting {}", file.toPath());
+            return new QuorumStateData();
+        }
+    }
+
+    /**
+     * Reads the election state from cached quorum data.
+     */
+    @Override
+    public ElectionState readElectionState() {
+        return new ElectionState(
+            cachedState.leaderEpoch(),
+            cachedState.leaderId() == UNKNOWN_LEADER_ID ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.leaderId()),
+            cachedState.votedId() == NOT_VOTED ? OptionalInt.empty() :
+            OptionalInt.of(cachedState.votedId()));
+    }
+
+    @Override
+    public void write(ElectionState latest) throws IOException {
+        cachedState.setLeaderEpoch(latest.epoch)
+            .setVotedId(latest.hasVoted() ? latest.votedId() : NOT_VOTED)
+            .setLeaderId(latest.hasLeader() ? latest.leaderId() : UNKNOWN_LEADER_ID);
+
+        final File temp = new File(stateFile.getAbsolutePath() + ".tmp");
```

Comment:
I guess we could leave this `.tmp` file behind in a hard crash? Should we check for it on startup and delete?

Reply:
Will delete upon write failure.

Reply:
I am not sure that is good enough. A hard crash in the middle of this operation could still leave the file around. I think we have to check for it on initialization.

Reply:
Sounds good